sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala

```diff
@@ -19,32 +19,69 @@ package org.apache.spark.sql.execution.datasources.v2
 import org.apache.spark.sql.connector.catalog.{StagingTableCatalog, TableCapability}
+import org.apache.spark.sql.connector.read.V1Scan
 import org.apache.spark.sql.connector.read.streaming.{ContinuousStream, MicroBatchStream}
-import org.apache.spark.sql.execution.{FilterExec, ProjectExec, SparkPlan}
+import org.apache.spark.sql.execution.{FilterExec, ProjectExec, RowDataSourceScanExec, SparkPlan}
 import org.apache.spark.sql.execution.datasources.DataSourceStrategy
 import org.apache.spark.sql.execution.streaming.continuous.{ContinuousCoalesceExec, WriteToContinuousDataSource, WriteToContinuousDataSourceExec}
+import org.apache.spark.sql.sources.TableScan
 import org.apache.spark.sql.util.CaseInsensitiveStringMap
 
-object DataSourceV2Strategy extends Strategy with PredicateHelper {
+class DataSourceV2Strategy(session: SparkSession) extends Strategy with PredicateHelper {
 
   import DataSourceV2Implicits._
 
   override def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
+    // projection and filters were already pushed down in the optimizer.
+    // this uses PhysicalOperation to get the projection and ensure that if the batch scan does
+    // not support columnar, a projection is added to convert the rows to UnsafeRow.
     case PhysicalOperation(project, filters, relation: DataSourceV2ScanRelation) =>
-      // projection and filters were already pushed down in the optimizer.
-      // this uses PhysicalOperation to get the projection and ensure that if the batch scan does
-      // not support columnar, a projection is added to convert the rows to UnsafeRow.
-      val batchExec = BatchScanExec(relation.output, relation.scan)
+      val output = relation.output
+      val pushedFilters = relation.getTagValue(V2ScanRelationPushDown.PUSHED_FILTERS_TAG)
+        .getOrElse(Seq.empty)
+
+      val (scanExec, needsUnsafeConversion) = relation.scan match {
+        case v1Scan: V1Scan =>
+          val v1Relation = v1Scan.toV1Relation(session.sqlContext)
+          if (v1Relation.schema != v1Scan.readSchema()) {
+            throw new IllegalArgumentException(
+              "The fallback v1 relation reports inconsistent schema:\n" +
+                "Schema of v2 scan:     " + v1Scan.readSchema() + "\n" +
+                "Schema of v1 relation: " + v1Relation.schema)
+          }
+          val rdd = v1Relation match {
+            case s: TableScan => s.buildScan()
+            case _ =>
+              throw new IllegalArgumentException(
+                "`V1Scan.toV1Relation` must return a `TableScan` instance.")
```

Comment:
If it must return `TableScan`, why not rename the API as `toV1TableScan` ?

Reply:
Sorry I meant rename to `toV1TableScan` and also change the return type as `TableScan`.

Reply:
`TableScan` is just a mixin, what we expect is `BaseScan with TableScan`, but that doesn't work for java.